File: javascript/syntax/5.arrays.js
// Arrays
let arr = ["apple", "banana", "cherry", "berry", "carry"];

console.log(arr[0]); // apple

// Length
console.log(arr.length); // 5

// Add item, adds to the end
arr.push("pineapple");
console.log(arr); // ['apple', 'banana', 'cherry', 'berry', 'carry', 'pineapple']

// Insert at index, adds "potatoe" at index 2 , changes original array
arr.splice(2, 0, "potatoe");
console.log(arr); // ['apple', 'banana', 'potatoe', 'cherry', 'berry', 'carry', 'pineapple']

// Extend/merge, spread operator
const arr2 = ["kiwi", "orange"];
arr.push(...arr2);
console.log(arr);
// ['apple', 'banana', 'potatoe', 'cherry', 'berry', 'carry', 'pineapple', 'kiwi', 'orange']

// Find index, returns -1 if not found
console.log(arr.indexOf("cherry")); // 3

// Check existence, returns boolean
console.log(arr.includes("banana")); // true

// find item, returns first match item
console.log(arr.find(element => element.startsWith("b"))); // banana

// Remove item, returns removed item
const appleIndex = arr.indexOf("apple");
if (appleIndex > -1) {
    arr.splice(appleIndex, 1);
}
console.log(arr);
// ['banana', 'potatoe', 'cherry', 'berry', 'carry', 'pineapple', 'kiwi', 'orange']

// Remove at index, removes 1 item at index 0, returns removed item, changes original array
arr.splice(0, 1);
console.log(arr); // ['potatoe', 'cherry', 'berry', 'carry', 'pineapple', 'kiwi', 'orange']

// Sort array
arr.sort();
console.log(arr); // ['berry', 'carry', 'cherry', 'kiwi', 'orange', 'pineapple', 'potatoe']
arr.sort((a, b) => b.localeCompare(a));
console.log(arr); // ['potatoe', 'pineapple', 'orange', 'kiwi', 'cherry', 'carry', 'berry']

// Reverse
arr.reverse();
console.log(arr); // ['berry', 'carry', 'cherry', 'kiwi', 'orange', 'pineapple', 'potatoe']

// Dynamic Array
const dynamicArr = new Array(5).fill(1);
console.log(dynamicArr);

// Map an array
const numArr = [1, 2, 3, 4, 5];
const squaredArr = numArr.map(x => x ** 2);
console.log(squaredArr); // [1, 4, 9, 16, 25]

// Filter an array
const filteredArr = numArr.filter(x => x % 2 === 0);
console.log(filteredArr); // [2, 4]

// Reduce an array
const reducedValue = numArr.reduce((sum, val) => sum + val, 0);
console.log(reducedValue); // 15

// Destructure array
const destructureArr = ["apple", "banana"];
let [a, b] = destructureArr;
console.log(a, b); // apple banana

// Swap
[a, b] = [b, a];
console.log(a, b); // banana apple

// Slices
const sliceArr = ["apple", "banana", "cherry", "berry", "carry"];

// Slicing (start-included : end-excluded), returns new array
console.log(sliceArr.slice(0, 1)); // ['apple']
console.log(sliceArr.slice(1, 2)); // ['banana']

// slicing (start-included : n) , returns new array
console.log(sliceArr.slice(1)); // ['banana', 'cherry', 'berry', 'carry']

// slicing (0: end-excluded) ,returns new array
console.log(sliceArr.slice(0, 2)); // ['apple', 'banana']

// slicing-reverse (0: end-included), returns new array
console.log(sliceArr.slice(0, -1)); // ['apple', 'banana', 'cherry', 'berry']

// slicing-reverse (start-excluded : end-included), returns new array
console.log(sliceArr.slice(-3, -1)); // ['cherry', 'berry'] 

File: java/syntax/5.Arrays.java
// Arrays
import java.util.*;
import java.util.stream.Collectors;

public class Arrays {
    public static void main(String[] args) {
        List<String> arr = new ArrayList<>(Arrays.asList("apple", "banana", "cherry", "berry", "carry"));

        System.out.println(arr.get(0)); // apple

        // Length
        System.out.println(arr.size()); // 5

        // Add item
        arr.add("pineapple");
        System.out.println(arr); // [apple, banana, cherry, berry, carry, pineapple]

        // Insert at index
        arr.add(2, "potatoe");
        System.out.println(arr); // [apple, banana, potatoe, cherry, berry, carry, pineapple]

        // Extend/merge
        List<String> arr2 = Arrays.asList("kiwi", "orange");
        arr.addAll(arr2);
        System.out.println(arr);
        // [apple, banana, potatoe, cherry, berry, carry, pineapple, kiwi, orange]

        // Remove item
        arr.remove("apple");
        System.out.println(arr);
        // [banana, potatoe, cherry, berry, carry, pineapple, kiwi, orange]

        // Remove at index
        arr.remove(0);
        System.out.println(arr); // [potatoe, cherry, berry, carry, pineapple, kiwi, orange]

        // Sort list
        Collections.sort(arr);
        System.out.println(arr); // [berry, carry, cherry, kiwi, orange, pineapple, potatoe]
        
        Collections.sort(arr, Collections.reverseOrder());
        System.out.println(arr); // [potatoe, pineapple, orange, kiwi, cherry, carry, berry]

        // Reverse
        Collections.reverse(arr);
        System.out.println(arr); // [berry, carry, cherry, kiwi, orange, pineapple, potatoe]

        // Dynamic Array
        List<Integer> dynamicArr = Collections.nCopies(5, 1);
        System.out.println(dynamicArr);

        // Map an array
        List<Integer> numArr = Arrays.asList(1, 2, 3, 4, 5);
        List<Integer> squaredArr = numArr.stream()
                .map(x -> x * x)
                .collect(Collectors.toList());
        System.out.println(squaredArr); // [1, 4, 9, 16, 25]

        // Filter an array
        List<Integer> filteredArr = numArr.stream()
                .filter(x -> x % 2 == 0)
                .collect(Collectors.toList());
        System.out.println(filteredArr); // [2, 4]

        // Reduce an array
        int reducedValue = numArr.stream().reduce(0, Integer::sum);
        System.out.println(reducedValue); // 15

        // Destructure array
        List<String> destructureArr = Arrays.asList("apple", "banana");
        String a = destructureArr.get(0);
        String b = destructureArr.get(1);
        System.out.println(a + " " + b); // apple banana

        // Swap
        String temp = a;
        a = b;
        b = temp;
        System.out.println(a + " " + b); // banana apple

        // Slices
        List<String> sliceArr = new ArrayList<>(Arrays.asList("apple", "banana", "cherry", "berry", "carry"));

        // Slicing (start-included : end-excluded)
        System.out.println(sliceArr.subList(0, 1)); // [apple]
        System.out.println(sliceArr.subList(1, 2)); // [banana]

        // slicing (start-included : n)
        System.out.println(sliceArr.subList(1, sliceArr.size())); // [banana, cherry, berry, carry]

        // slicing (0: end-excluded)
        System.out.println(sliceArr.subList(0, 2)); // [apple, banana]

        // slicing-reverse (0: end-included)
        System.out.println(sliceArr.subList(0, sliceArr.size() - 1)); // [apple, banana, cherry, berry]

        // slicing-reverse (start-excluded : end-included)
        System.out.println(sliceArr.subList(sliceArr.size() - 3, sliceArr.size() - 1)); // [cherry, berry]
    }
} 